platform/sync-installer/sync-installer-impl/src/main/java/org/codice/ddf/sync/installer/impl/SynchronisedInstallerImpl.java

```diff
@@ -0,0 +1,566 @@
+      throws SynchronisedInstallerException {
+    return createManagedFactoryService(
+        DEFAULT_MAX_SERVICE_WAIT, factoryPid, properties, bundleLocation);
+  }
+
+  @Override
+  public Configuration createManagedFactoryService(
+      long maxWaitTime, String factoryPid, Map<String, Object> properties, String bundleLocation)
+      throws SynchronisedInstallerException {
+    LOGGER.info("Creating managed service of factorypid [{}]", factoryPid);
+
+    Configuration createdConfig;
+    try {
+      createdConfig = configAdmin.createFactoryConfiguration(factoryPid, null);
+      createdConfig.setBundleLocation(bundleLocation);
+    } catch (IOException e) {
+      throw new SynchronisedInstallerException(
+          "Failed to initialize managed service factory configuration with fpid of [" + factoryPid + "]",
+          e);
+    }
+
+    try {
+      if (properties == null || properties.isEmpty()) {
+        createdConfig.update();
+      } else {
+        createdConfig.update(toDic(properties));
+      }
+    } catch (IOException e) {
+      throw new SynchronisedInstallerException(
+          "Failed to update created managed service factory configuration with pid of ["
+              + createdConfig.getPid()
+              + "]",
+          e);
+    }
+
+    waitForServiceAvailability(maxWaitTime, createdConfig.getPid());
+    return createdConfig;
+  }
+
+  @Override
+  public void updateManagedService(
+      String servicePid, Map<String, Object> properties, String bundleLocation)
+      throws SynchronisedInstallerException {
+    updateManagedService(DEFAULT_MAX_SERVICE_WAIT, servicePid, properties, bundleLocation);
+  }
+
+  @Override
+  public void updateManagedService(
+      long maxWaitTime, String servicePid, Map<String, Object> properties, String bundleLocation)
+      throws SynchronisedInstallerException {
+    waitForServiceAvailability(maxWaitTime, servicePid);
+
+    LOGGER.debug(
+        "Registering service configuration listener to listen to service with pid of {}.",
+        servicePid);
+    ServiceConfigurationListener listener = new ServiceConfigurationListener(servicePid);
+    ServiceRegistration<?> registration =
+        bundleContext.registerService(ConfigurationListener.class.getName(), listener, null);
+
+    Callable<Boolean> isServiceUpdated =
+        () -> {
+          LOGGER.debug(
+              "Waiting for service with pid [{}] to reflect configuration updates...", servicePid);
+          return listener.isUpdated();
+        };
+
+    boolean available;
+
+    LOGGER.info("Updating configuration of service with pid [{}].", servicePid);
+
+    try {
+      Configuration config = configAdmin.getConfiguration(servicePid, bundleLocation);
+      config.setBundleLocation(bundleLocation);
+      config.update(toDic(properties));
+
+      LOGGER.debug("Updated configuration of service with pid [{}].", servicePid);
+      available = wait(isServiceUpdated, maxWaitTime, DEFAULT_POLLING_INTERVAL);
```

Comment:
❗️ At this point you should wait for whatever is left since you already waited a bit above. Otherwise, a client saying to wait for 5 seconds might end up waiting for 10 seconds.